### ISSUE TITLE
Clarify the restriction for `minValue` and `maxValue` of `MLClampOptions`

### ISSUE THREAD

**Issue by huningxin** [webmachinelearning/webnn#396]
WebNN [`clamp`](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-clamp) limits the input tensor element-wise within a range specified by the minimum and maximum values. The minimum and maximum values could be specified by `MLClampOptions`.

This open item was raised by @quidity (Thanks Alex!) in Chromium [CL review](https://chromium-review.googlesource.com/c/chromium/src/+/4597587/comment/d1368beb_9079fbb8/) which is

> can min & max be == floats and is that ok or not?

The current [Chromium implementation](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/ml/webnn/ml_graph_builder.cc;l=70)  restricts the min value should be less than or equal to the max value. TensorFlow.js also implements [this restriction](https://github.com/tensorflow/tfjs/blob/tfjs-v4.7.0/tfjs-core/src/ops/clip_by_value.ts#L47). However, some backends implement more strict restriction that the min value should be less than max value, like [XNNPACK](https://github.com/google/XNNPACK/blob/master/src/operators/unary-elementwise-nc.c#L328).

WebNN spec should clarify the restriction.

/cc @fdwr, for inputs on DirectML .

**Comment by fdwr** [comment]
> WebNN spec should clarify the restriction.

We have a few options, but I think we should:

- a. Not exclude infinities. Positive infinity should be a perfectly legal `maximum`, and negative infinity a legal `minimum`.
- b. Consider rather than rejecting inverted ranges at the WebNN API layer, set min/max explicitly to whatever behavior we want before passing the values to the lower layer backends. Though, if we're not able to decide, then rejecting at the WebNN API layer and forcing the caller to massage min/max values before calling WebNN is reasonable. Since `clamp` only accepts scalar min/max values, it's a much easier situation to handle than if clip accepted entire *tensors* for min/max.

# Framework comparisons

Given `[0,5,10,15]` and the inverted range `min=10` and `max=5`:

- TensorFlow: 10, 10, 10, 10
- PyTorch: 5, 5, 5, 5
- C++ `std::clamp`: 10, 10, 5, 5
- DirectML: 10, 10, 10, 10

As you can see, there's far from universal consensus on how to treat inverted ranges 😅🙃.

# TensorFlow

https://www.tensorflow.org/api_docs/python/tf/clip_by_value

```
import tensorflow as tf

x = tf.constant([0,5,10,15], dtype=tf.float32)
y = tf.clip_by_value(x, 10, 5); # https://www.tensorflow.org/api_docs/python/tf/clip_by_value
print("value:", y)
print("shape:", y.shape)

# Prints:
# value: tf.Tensor([10. 10. 10. 10.], shape=(4,), dtype=float32)
# shape: (4,)
```

# PyTorch

https://pytorch.org/docs/stable/generated/torch.clamp.html

```
import torch

x = torch.tensor([0,5,10,15], dtype=torch.float)
y = torch.clamp(x, min=10, max=5)

print("value:", y)
print("shape:", y.shape)
# Prints:
# value: tensor([5., 5., 5., 5.])
# shape: torch.Size([4])
```

# ONNX

https://github.com/onnx/onnx/blob/main/docs/Operators.md#Clip

ONNX [Clip](https://github.com/onnx/onnx/blob/main/docs/Operators.md#Clip) doesn't specify.

# C++

https://en.cppreference.com/w/cpp/algorithm/clamp

C++ [std::clamp](https://en.cppreference.com/w/cpp/algorithm/clamp):

```
if x < min, then min
if x > max, then max
else x
```

```
#include <iostream>
#include <algorithm>

int main()
{
    int x;
    std::cout << "Given min=10 max=5" << std::endl;
    std::cout << (x = 0 ) << " -> " << std::clamp(x, 10, 5) << std::endl;
    std::cout << (x = 5 ) << " -> " << std::clamp(x, 10, 5) << std::endl;
    std::cout << (x = 10) << " -> " << std::clamp(x, 10, 5) << std::endl;
    std::cout << (x = 15) << " -> " << std::clamp(x, 10, 5) << std::endl;

    return 0;
}
```
Prints:

```
Given min=10 max=5
0 -> 10
5 -> 10
10 -> 5
15 -> 5
```

# DirectML

For [DML_OPERATOR_ELEMENT_WISE_CLIP](https://learn.microsoft.com/en-us/windows/win32/api/directml/ns-directml-dml_element_wise_clip_operator_desc), it's `max(minValue, min(x, maxValue))`. *Note the doc currently says `f(x) = min(Max, max(Min, x))`, but this is inconsistent with our output - I reached out to our doc writer to fix it.

```
dxdispatch.exe models\dml_element_wise_clip.json
Running on 'NVIDIA Quadro P620'
Dispatch 'clip': 1 iterations, 1.0926 ms median (CPU), 0.0051 ms median (GPU)
Resource 'A': 10, 10, 10, 10
```

```
{
    "$schema": "./_schema.json",
    
    "resources": 
    {
        "A": 
        {
            "initialValuesDataType": "FLOAT32",
            "initialValues": [0,5,10,15]
        }
    },

    "dispatchables": 
    {
        "clip": 
        {
            "type": "DML_OPERATOR_ELEMENT_WISE_CLIP",
            "desc": 
            {
                "InputTensor": { "DataType": "FLOAT32", "Sizes": [4] },
                "OutputTensor": { "DataType": "FLOAT32", "Sizes": [4] },
                "Min": 10,
                "Max": 5
            }
        }
    },

    "commands": 
    [
        {
            "type": "dispatch",
            "dispatchable": "clip",
            "bindings": 
            {
                "InputTensor": "A",
                "OutputTensor": "A"
            }
        },
        { "type": "print", "resource": "A" }
    ]
}
```

**Comment by fdwr** [comment]
p.s. Unlike the inverted case `min > max` which is arguable about what the result should be, the `min == max` case is unambiguous and consistent across each of these. That XNNPack rejects min == max 😳 feels like a bug typo.

**Comment by wchao1115** [comment]
As a backend API, I think WebNN should not allow a corner case situation that makes it hard for a specific implementation to follow thru e.g. if WebNN supports a degenerate range min == max while XNNPACK can't implement them, that is a burden on the implementer, one that could lead to a corner case latent bug that no one knows about until it hits a customer. Even for the implementation that does support it, it is still likely that they may support them differently depending on their policy or ambiguous circumstances e.g. as a no-op, a min filling, etc. 

Therefore, in this situation I think it is better to assert that min < max, and not min <= max. If the framework on top wants to support a degenerate range, it should be pretty easy for them to assert their own policy there. 

As to whether we should also support inverted range, my rationale is similar to the degenerate range case, it is non-obvious and supporting non-obvious case could lead to a bug-compatibility burden on the implementation side in the future.

**Comment by huningxin** [comment]
XNNPACK allows min and max to be the same: https://github.com/google/XNNPACK/commit/d98c1ff25f7872ff91d5179c7f9374c6ee95df08 (Thanks @fdwr for the notice)

**Comment by inexorabletash** [comment]
> **tldr**: All other tested ML libraries handle empty ranges properly as expected (e.g. `min=42`, `max=42`). So my opinion is that we should open an issue to fix the outlier, by removing that suspiciously-probably-a-typo [validation check](https://github.com/tensorflow/tfjs/blob/tfjs-v4.7.0/tfjs-core/src/ops/clip_by_value.ts#L47) in TF.js. Empty ranges are not an ambiguous corner case:

Hey @fdwr - is there confusion about TF.js here or am I reading your comment wrong? TF's test (`util.assert((clipValueMin <= clipValueMax),...`) allows empty ranges (min == max). I'm not seeing a glitch in what you linked.

I believe XNNPACK was the only outlier and that's been fixed, so I think we're good to resolve this and remove the note in the spec referencing this issue. Can you confirm, or clarify the remaining concern?

**Comment by fdwr** [comment]
@inexorabletash Indeed, TF.js looks good too. So that's all of them then, and the note should be removed. Updated the table above.
